Realm/RLMRealm.mm

```diff
@@ -381,8 +378,69 @@ + (instancetype)realmWithPath:(NSString *)path
     }
 
     key = key ?: keyForPath(path);
-    realm = [[RLMRealm alloc] initWithPath:path key:key readOnly:readonly inMemory:inMemory dynamic:dynamic error:outError];
-    if (outError && *outError) {
+    if ((key = validatedKey(key))) {
+        validateNotInDebugger();
+    }
+
+    realm = [RLMRealm new];
+    realm->_dynamic = dynamic;
+
+    realm::Realm::Config config;
+    config.path = path.UTF8String;
+    config.read_only = readonly;
+    config.in_memory = inMemory;
+    config.encryption_key = key ? static_cast<const char *>(key.bytes) : StringData();
+
+    __weak RLMRealm *weakRealm = realm;
+    config.migration_function = [=](__unused Group *group, ObjectStore::Schema &schema) {
+        RLMRealm *strongRealm = weakRealm;
+        if (!strongRealm) {
+            return;
+        }
```

Comment:
I'm unsure if this is guaranteed to work. Underneath, the C++ lambda gets turned into a struct with a `__weak RLMRealm*` member, but I'm unsure if the `__weak` qualifier is actually applied to that member, or if it even can be applied to it. Does `__weak` generally work for struct members, or only for stack variables?